From d55456156b197614c18d57801e6d6e14c8e0c071 Mon Sep 17 00:00:00 2001
From: Sam James <sam@gentoo.org>
Date: Sat, 10 Sep 2022 11:20:27 +0100
Subject: [PATCH] autotools

--- /dev/null
+++ b/Makefile.am
@@ -0,0 +1 @@
+info_TEXINFOS = c.texi
--- /dev/null
+++ b/configure.ac
@@ -0,0 +1,7 @@
+AC_PREREQ([2.71])
+AC_INIT([c-intro-and-ref], [0], [c-manual@gnu.org])
+
+AM_INIT_AUTOMAKE([-Wall -Werror foreign])
+AC_CONFIG_FILES([Makefile])
+AC_OUTPUT
+
