diff --git a/src/ninja.cc b/src/ninja.cc
index d763d4f..ff2b57d 100644
--- a/src/ninja.cc
+++ b/src/ninja.cc
@@ -1633,10 +1633,6 @@ bool JobserverState::ShouldSetupClient(const BuildConfig& config,
     *reason = "Dry-run mode";
     return false;
   }
-  if (config.explicit_parallelism && !config.jobserver_pool) {
-    *reason = "Explicit parallelism specified";
-    return false;
-  }
   return HasExternalJobserverPool(reason);
 }
 
