--- a/src/CMakeLists.txt 2020-04-13 16:02:49.754007602 -0000
+++ a/src/CMakeLists.txt 2020-04-13 16:03:33.931709811 -0000
@@ -1,5 +1,7 @@
 cmake_minimum_required(VERSION 3.1.0)

+include(GNUInstallDirs)
+
 find_package(StdFileSystem)

 if (BUILD_SHARED)
@@ -17,7 +19,7 @@
   install(
     TARGETS cppglob
     EXPORT cppglob-config
-    DESTINATION lib
+    DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}
     COMPONENT libs)

 endif()
